Bulk Removal of submissions via admin search

vkurup opened the pull request:
This:
- adds a `list_filter` to make it easier for admins to only start with 'approved' submissions
- adds a `search_fields` setting to allow admins to search for any text in the `idea` column. (Note, the headline and followup fields are merged into the idea field during normal (non-admin) form submission)
- adds a custom action to allow 'removal' of submissions chosen via checkbox (after a confirmation page)
- emails the original user about the change, but only if it really was a change. If original submission was already unapproved, no email is sent.
- is based on code stolen from Django's `delete_selected` action.

My choice of the label for the dropdown is "Remove selected submissions", which is confusingly similar to "Delete selected submissions", but I couldn't think of a better, not-too-long name. Ideas? Or maybe we should remove the `delete_selected` action to prevent confusion?

vkurup committed:
Bulk Removal of submissions via admin search

vkurup commented:
Hmmm... not sure what's causing those build failures (and why it succeeded on the PR, but not the push). Any ideas? If not, will investigate more tomorrow.

ejucovy commented on opendebates/admin.py:
```diff
@@ -15,9 +17,33 @@ class CategoryAdmin(ModelAdmin):
+                submission.approved = False
+                submission.moderated_removal = True
+                submission.removal_flags.all().update(reviewed=True)
+                submission.save()
```
Should we be worried about the 2N `update`s this will generate?  I guess N will never be more than 100, and probably closer to 5-20 at a time in normal usage.  But 10-40 queries still seems like a lot.  Is it worth making this more like 3 queries instead of 2N?  (One would get the count or maybe a list of PKs, one would be a `queryset.filter(approved=True).update(..)`, and the third would be something like `Flag.objects.filter(to_remove__in=..).update(..)`)

vkurup replied:
Yes, I actually wrote a version with just those queries (and the first one needs to get a list of PKs so we can email later), but then reverted it to this because I thought this was clearer to read :) I didn't think a couple hundred queries was a problem since it's an admin action and not likely to happen frequently, but definitely see your point. Jeff, any preference? (he's taken over this from me)

jbradberry replied:
I don't think it's too many queries, but I also don't think changing it to be a pair of updates would be a bad idea, and generally I lean towards efficiency.

jbradberry replied:
On second thought, since we'll have to iterate one at a time anyway to do the emails, I'm inclined to leave it as-is.

ejucovy commented on opendebates/admin.py:
```diff
@@ -15,9 +17,33 @@ class CategoryAdmin(ModelAdmin):
+                submission.moderated_removal = True
+                submission.removal_flags.all().update(reviewed=True)
+                submission.save()
+                send_email("idea_is_removed", {"idea": submission})
```
Similarly - is there any reason to worry about generating so many little celery tasks all at once in a single request?  (I have even less intuition here than with the db queries.)  We could also write a `send_emails("idea_is_removed", array_of_ctxs)` utility that does a single longer running task; or would that be worse?

vkurup replied:
I'd think that celery should be ideally suited to handle this, and quicker smaller tasks would be better than a longer one.

jbradberry replied:
Yeah, this shouldn't be an issue, I'd think.

ejucovy commented on opendebates/admin.py:
```diff
@@ -15,9 +17,33 @@ class CategoryAdmin(ModelAdmin):
+            return None  # returning None causes us to return to changelist
+        context = {'queryset': queryset}
+        return render(request, 'opendebates/admin/remove_submissions.html', context)
+    remove_submissions.short_description = 'Remove selected submissions'
```
Maybe "Unmoderate selected submissions and notify submitters"?  Or "Unapprove" if that's more consistent with language elsewhere?  

(I think we want to remind the moderators that this will trigger emails, even if it makes the label really long.)

Getting rid of the delete option also makes sense to me, I can't see why it would ever be needed/wanted.

jbradberry replied:
Here I like the 'remove' language since it matches the name/url of the remove moderation view.  Also, we have this set up as a two-step process, where the details of what happens are displayed for the reviewer, so I think the short description is ok here.

ejucovy replied:
Now I'm thinking we should get rid of the built-in delete option.  The second step of the process looks & feels broadly similar in both actions, so they would actually need to attend closely to the text on the second screen to know which one they're being prompted to confirm.  And I'm just worried that with moderators being so overwhelmed, they're going to stop reading that confirmation page closely sooner or later.

jbradberry replied:
Agreed.  Removing.

ejucovy reviewed (commented):
Re: the build failures, running `manage.py collectstatic` to populate the static directory before the test run appears to fix it, for what that's worth.  But actually, if I follow both builds' steps verbatim on a previously empty directory, they _both_ fail for me in the first run.  Is there any way that the PR one might not be starting from a completely clean server?  (I don't really know how travis works.)

jbradberry commented on opendebates/admin.py:
```diff
@@ -15,9 +17,33 @@ class CategoryAdmin(ModelAdmin):
+        if request.POST.get('post'):
+            count = 0
+            # only email user if submission changes status, hence the filter
+            for submission in queryset.filter(approved=True):
```
This will only be useful if the questions have already been flagged.  If the moderation team is running around having to flag everything themselves, we are also going to need a bulk flag action, I think.

jbradberry replied:
... or perhaps I misunderstand what is meant by `approved=True` here.  That's also possible.

jbradberry replied:
I clearly misunderstood.  Never mind ;)

vkurup replied:
Yeah... submissions are 'approved' by default, and only unapproved through moderation.

jbradberry committed:
Removed the 'delete selected' action from SubmissionAdmin.

It's too similar to 'remove', and we want to keep the removed
questions around just in case.

jbradberry committed:
Changed the static files settings when tests are being run.

jbradberry commented:
@ejucovy Got it working, just a couple of staticfile settings were inappropriate while tests are being run.  Are you good with this as-is?

ejucovy commented:
Yup!
